src/nanotron/models/llama.py

```diff
@@ -46,6 +53,23 @@
 logger = logging.get_logger(__name__)
 
 
+class RMSNorm(nn.Module):
+    def __init__(self, dim: int, eps: float = 1e-6):
+        super().__init__()
+        self.eps = eps
+        self.weight = nn.Parameter(torch.zeros(dim))
+
+    def _norm(self, x):
+        return x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + self.eps)
+
+    def forward(self, x):
+        output = self._norm(x.float())
+        # Llama does x.to(float16) * w whilst Gemma is (x * w).to(float16)
+        # See https://github.com/huggingface/transformers/pull/29402
+        output = output * (1.0 + self.weight.float())
+        return output.type_as(x)
+
+
 class RotaryEmbedding(nn.Module):
```

Comment:
Didn't get it. I didn't change the rotary embedding.

Reply:
Sorry if it wasn't clear. I mean like instead of putting RotaryEmbedding for inference and training in one class, we should split it to RotaryEmbeddingForInference, and RotaryEmbeddingForTraining, then based on the config, we call the corresponding one in CausalSelfAttention.